### PR TITLE
eof: Add trait since when an instruction is in EOF

### PR DESCRIPTION
**Pull request by chfast**
Add `std::optional<evmc_revision> eof_since` to the `instr::traits` with the information in which EVM revision an instruction has become valid in EOF.

Using `std::optional` has some inconvenience:
the `std::nullopt` is less-than any concrete value (think "Frontier-1") while for undefined instructions we rather want to assign +infinity.

### CODE REVIEW

File: lib/evmone/instructions_traits.hpp [hunk 1]
```diff
@@ -200,6 +200,9 @@ constexpr inline GasCostTable gas_costs = []() noexcept {
 static_assert(gas_costs[EVMC_MAX_REVISION][OP_ADD] > 0, "gas costs missing for a revision");
 
 
+constexpr auto REV_EOF1 = EVMC_PRAGUE;
```
Comment (chfast) [inline lib/evmone/instructions_traits.hpp:203]: Good idea?

File: lib/evmone/instructions_traits.hpp [hunk 2]
```diff
@@ -223,6 +226,8 @@ struct Traits
     /// every EVM revision the value is ::EVMC_FRONTIER. For undefined instructions the value is not
     /// available.
     std::optional<evmc_revision> since;
+
+    std::optional<evmc_revision> eof_since;
```
Comment (chfast) [inline lib/evmone/instructions_traits.hpp:230]: Comment missing.

**Comment by gumb0** [comment]
Another idea is having one `since` revision + 2 flags `legacy` and `eof`, assuming when instruction is activated in both, it's always activated in the same revision.

Argument against can be that it can be seen as some old opcodes are activated before Prague, but also have `eof` flag, so might be interpreted as "activated in eof before Prague", but I think it shouldn't influnce execution, it's equivalent to the table before this change.

**Comment by chfast** [comment]
> Another idea is having one `since` revision + 2 flags `legacy` and `eof`, assuming when instruction is activated in both, it's always activated in the same revision.
> 
> Argument against can be that it can be seen as some old opcodes are activated before Prague, but also have `eof` flag, so might be interpreted as "activated in eof before Prague", but I think it shouldn't influnce execution, it's equivalent to the table before this change.

I considered this but this is less generic, requires one more field, and the main usage (build cost table) would need to use both revision+flag anyway.

File: lib/evmone/instructions_traits.hpp (hunk 1 above)
Comment (chfast) [reply lib/evmone/instructions_traits.hpp:203]: Is the name good? For `TXCREATE` we need something like "EOF v1.1".

Comment (gumb0) [reply lib/evmone/instructions_traits.hpp:203]: I think it's okay. Also okay to drop `TXCREATE`

**Review by pdobacz — APPROVED** [review]
lgtm, will there also be `eof_until`, in case we ever start removing instructions from subsequent EOF majors?

File: lib/evmone/instructions_traits.hpp [hunk 3]
```diff
@@ -223,6 +226,10 @@ struct Traits
     /// every EVM revision the value is ::EVMC_FRONTIER. For undefined instructions the value is not
     /// available.
     std::optional<evmc_revision> since;
+
+    /// The EVM revision in which the instruction has been valid in EOF. For instructions invalid
```
Comment (gumb0) [inline lib/evmone/instructions_traits.hpp:230]:
```suggestion
    /// The EVM revision in which the instruction has become valid in EOF. For instructions invalid
```

**Review by gumb0 — APPROVED** [review]

**Pull request merged**